### PR TITLE
feature: propagate component status codes to FSS

### PR DESCRIPTION
**Pull request by indougnito**
**Issue #, if available:**

**Description of changes:**

Propagates component status codes to FSS.

Implements a new lifecycle method reportState(newState, statusCode) accepting a status code parameter, which will be propagated to FSS in a fashion similar to lifecycle state.

Invoking reportState(newState) without a status code parameter causes a default status code to be propagated, e.g. ComponentStatusCode.RUN_ERRORED if the new state is ERRORED and the previous state was RUNNING, or ComponentStatusCode.NONE if the new state is healthy.

Component status updates will now be triggered on ERRORED in addition to the existing BROKEN state transitions 

Also updates the config README to call attention to some Topic naming details.

**Why is this change necessary:**
We would like to know why a component failed, i.e. its status code, in addition to the fact that it failed.

**How was this change tested:**
- [x] Updated or added new unit tests.
- [x] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [x] Updated the README if applicable.

**Compatibility Checklist:**
- [x] I confirm that the change is backwards compatible.
- [x] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/lifecyclemanager/Lifecycle.java [hunk 1]
```diff
@@ -223,6 +278,12 @@ private Topic initStateTopic(final Topics topics) {
         return state;
     }
 
+    private Topic initTopic(final Topics topics, final String topicName) {
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/lifecyclemanager/Lifecycle.java:281]: nit: initStateTopic should be rewritten to call this

File: src/main/java/com/aws/greengrass/lifecyclemanager/Lifecycle.java [hunk 2]
```diff
@@ -265,11 +326,11 @@ private void setDesiredState(State... state) {
...
             // try insert to the queue, if queue full doesn't block.
-            enqueueStateEvent("DesiredStateUpdated");
+            enqueueStateEvent(new DesiredStateUpdatedEvent());
         }
     }
 
-    private void enqueueStateEvent(Object event) {
+    private void enqueueStateEvent(StateEvent event) {
         if (!stateEventQueue.offer(event)) {
             logger.error("couldn't put the new event to stateEventQueue");
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/lifecyclemanager/Lifecycle.java:335]: nit: should the event be logged here?

File: src/main/java/com/aws/greengrass/status/FleetStatusService.java [hunk 3]
```diff
@@ -295,6 +295,21 @@ private void handleServiceStateChange(GreengrassService greengrassService, State
             updatedGreengrassServiceSet.add(greengrassService);
         }
 
+        // Always report status on errored state, evaluate overall status based on current state of all components
+        if (newState.equals(State.ERRORED)) {
+            synchronized (updatedGreengrassServiceSet) {
+                Instant now = Instant.now();
+                AtomicReference<OverallStatus> overAllStatus = new AtomicReference<>();
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/status/FleetStatusService.java:302]: we've seen some circumstances where overallStatus is null - is it worth initializing it to HEALTHY?

**Review by rbattle — APPROVED** [review]
looks ok - few minor comments

File: src/main/java/com/aws/greengrass/status/FleetStatusService.java (hunk 3 above)
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/status/FleetStatusService.java:302]: overallStatus would be null if kernel just launches and there is no service running. Should we consider it healthy or assign a new value like "UNKNOWN"?

(I think we can address this issue in a separate pr.

File: src/main/java/com/aws/greengrass/lifecyclemanager/GenericExternalService.java [hunk 4]
```diff
@@ -807,4 +810,12 @@ protected Exec addUser(Exec exec, boolean requiresPrivilege) {
         }
         return exec;
     }
+
+    @AllArgsConstructor
+    @Data
+    private static class RunResult {
+        private RunStatus runStatus;
+        private Exec exec;
+        private ComponentStatusCode statusCode;
```
[the remark this answers is not in the document]
Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/lifecyclemanager/GenericExternalService.java:819]: Could you also add integ test for an invalid runwith or skipif config to make sure that that path works?

**Review by junfuchen99 — DISMISSED** [review]

**Review by junfuchen99 — DISMISSED** [review]

**Review by junfuchen99 — APPROVED** [review]

**Review by kucnavya — APPROVED** [review]

**Pull request merged**